tests/flow_control.rs

```diff
@@ -224,3 +225,206 @@ fn recv_data_overflows_window() {
 fn recv_window_update_causes_overflow() {
     // A received window update causes the window to overflow.
 }
+
+#[test]
+fn stream_close_by_data_frame_releases_capacity() {
+    let _ = ::env_logger::init();
+    let (m, mock) = mock::new();
+
+    let window_size = frame::DEFAULT_INITIAL_WINDOW_SIZE as usize;
+
+    let h2 = Client::handshake(m).unwrap()
+        .and_then(|mut h2| {
+            let request = Request::builder()
+                .method(Method::POST)
+                .uri("https://http2.akamai.com/")
+                .body(()).unwrap();
+
+            // Send request
+            let mut s1 = h2.request(request, false).unwrap();
+
+            // This effectively reserves the entire connection window
+            s1.reserve_capacity(window_size);
+
+            // The capacity should be immediately available as nothing else is
+            // happening on the stream.
+            assert_eq!(s1.capacity(), window_size);
+
+            let request = Request::builder()
+                .method(Method::POST)
+                .uri("https://http2.akamai.com/")
+                .body(()).unwrap();
+
+            // Create a second stream
+            let mut s2 = h2.request(request, false).unwrap();
+
+            // Request capacity
+            s2.reserve_capacity(5);
+
+            // There should be no available capacity (as it is being held up by
+            // the previous stream
+            assert_eq!(s2.capacity(), 0);
+
+            // Closing the previous stream by sending an empty data frame will
+            // release the capacity to s2
+            s1.send_data("".into(), true).unwrap();
+
+            // The capacity should be available
+            assert_eq!(s2.capacity(), 5);
+
+            // Send the frame
+            s2.send_data("hello".into(), true).unwrap();
+
+            // Wait for the connection to close
+            h2.unwrap()
+        })
+        ;
+
+    let mock = mock.assert_client_handshake().unwrap()
+        // Get the first frame
```

Comment:
Just to illustrate the changes in #64, this could be written:

```rust
let mock = mock.assert_client_handshake().unwrap()
    .recv_settings()
    .recv_frame(
        frames::headers(1)
            .request("POST", "https://http2.akamai.com")
    )
    .recv_frame(
        frames::headers(3)
            .request("POST", "https://http2.akamai.com")
    )
    .recv_frame(
        frames:data(1, "")
            .eos()
    )
    .recv_frame(
        frames:data(3, "hello")
            .eos()
    )
    .map(drop)
```

Reply:
Feel free to merge #64 and update the test 👍